src/systemd-sonic-generator/ssg-test.cc

```diff
@@ -0,0 +1,550 @@
+        asic_conf_format_ = g_asic_conf_format;
+    }
+
+    /* Restore global vars */
+    virtual void TearDown() {
+        g_unit_file_prefix = unit_file_prefix_;
+        g_config_file = config_file_;
+        g_machine_config_file = machine_config_file_;
+        g_asic_conf_format = asic_conf_format_;
+
+        g_ssg_test_mutex.unlock();
+    }
+
+  private:
+    const char* unit_file_prefix_;
+    const char* config_file_;
+    const char* machine_config_file_;
+    const char* asic_conf_format_;
+};
+
+/* 
+ * class SsgFunctionTest
+ * Implements functions to execute functional level tests.
+ */
+class SsgFunctionTest: public SystemdSonicGeneratorFixture {
+  protected:
+    /* This function generates the generated_services.conf file */
+    void generate_generated_services_conf() {
+        FILE* fp = fopen(TEST_CONFIG_FILE.c_str(), "w");
+        ASSERT_NE(fp, nullptr);
+        for (std::string str : generated_services) {
+            fputs(str.c_str(), fp);
+            fputs("\n", fp);
+        }
+        fclose(fp);
+    }
+
+    /* copy files from src_dir to dest_dir */
+    void copyfiles(const char* src_dir, const char* dest_dir) {
+        // Iterate through the source directory
+        for (fs::directory_iterator file(src_dir);
+                 file != fs::directory_iterator(); ++file) {
+            try {
+                fs::path current(file->path());
+                if(!fs::is_directory(current)) {
+                    /* Copy file */
+                    fs::copy_file( current, dest_dir / current.filename());
+                }
+            }
+            catch(fs::filesystem_error const & e) {
+                std:: cerr << e.what() << '\n';
+            }
+        }
+    }
+
+    /* Save global variables before running tests */
+    virtual void SetUp() {
+        FILE* fp;
+        SystemdSonicGeneratorFixture::SetUp();
+
+        /* Setup Input and Output directories and files */ 
+        fs::path path{TEST_UNIT_FILE_PREFIX.c_str()};
+        fs::create_directories(path);
+        path = fs::path(TEST_OUTPUT_DIR.c_str());
+        fs::create_directories(path);
+        path = fs::path(TEST_PLATFORM_DIR.c_str());
+        fs::create_directories(path);
+        fp = fopen(TEST_MACHINE_CONF.c_str(), "w");
+        ASSERT_NE(fp, nullptr);
+        fputs("onie_platform=test_platform", fp);
+        fclose(fp);
+        generate_generated_services_conf();
+        copyfiles(TEST_UNIT_FILES.c_str(), TEST_UNIT_FILE_PREFIX.c_str());
+    }
+
+    /* Restore global vars */
+    virtual void TearDown() {
+        /* Delete ssg_test directory */
+        EXPECT_TRUE(fs::exists(TEST_ROOT_DIR.c_str()));
+        fs::path path{TEST_ROOT_DIR.c_str()};
+        fs::remove_all(path);
+
+        SystemdSonicGeneratorFixture::TearDown();
+    }
+
+  private:
+};
+
+/* 
+ * class SsgMainTest
+ * Implements functions to test ssg_main routine.
+ */
+class SsgMainTest: public SsgFunctionTest {
```

Comment:
Suggest to add a space before first `:` to align style with other SONiC C++ files.